### PR TITLE
Cmfd refactor to sync with ClosedMOC/3D-MOC

### PR DESCRIPTION
**Pull request by samuelshaner**
This PR will be accompanied by a PR to ClosedMOC/3D-MOC that syncs up the Cmfd, Vector, and Matrix classes and linalg functions as much as possible. Important changes in this PR include:

* **Shared OpenMP locks**: @geogunow suggested that OpenMP locks be shared across objects, where possible, to avoid unnecessary duplication of locks. Cmfd now has an array of locks (by Cmfd cell) that are shared with the Vector and Matrix objects used to solve the diffusion eigenvalue problem.
* **splitEdgeCurrents() corrected**: The procedure for splitting edge currents has been fixed to properly handle REFLECTIVE, PERIODIC, and VACUUM boundary conditions.
* **Solver residual initialized to 0.**: If left uninitialized, the residual output for iteration 0 is junk (at least for double precision).
* **Vectorized `setValues(...)` and `incrementValues(...)` added to Vector.cpp**: This implements the Cmfd shared memory optimization that @jtramm started in PR #196.
* **`Lattice::getLatticeSurface(...)` reverted to general form**: The `Lattice::getLatticeSurface(...)` method made some assumptions on the numbering of the surfaces and edges that is not valid for 3D surfaces, so this method was reverted to the previous general form.
* **`splitCorners()` -> `splitEdgeCurrents()`**: The `splitCorners()` routine has been changed to `splitEdgeCurrents()` and the nomenclature for corners has been changed to edges to reflect the language commonly used elsewhere to describe a cube:

faces-edges-vertices

### CODE REVIEW

**Comment by wbinventor** [comment]
I'll make a quick review if this after @geogunow does primary thorough review, since I've been using CMFD a great deal. I'm curious, do these.changes affect the results, e.g., the CMFD test results from #251?

**Comment by samuelshaner** [comment]
I just ran a bunch of cases from the sample-inputs, including `c5g7-cmfd.py`, and the `test_cmfd_pwr_assembly` and got the exact same results from the code in this PR and the code from PR #251.

**Comment by wbinventor** [comment]
Great! Thanks for checking @samuelshaner.

File: src/Cmfd.cpp [hunk 1]
```diff
@@ -1087,57 +1067,64 @@ void Cmfd::setNumFSRs(int num_fsrs) {
...
-  log_printf(INFO, "Splitting CMFD corner currents...");
+  log_printf(INFO, "Splitting CMFD edge currents...");
 
   int ncg = _num_cmfd_groups;
+  int nf = NUM_FACES;
+  int ne = NUM_EDGES;
+  int ns = NUM_SURFACES;
 
-  #pragma omp parallel
+#pragma omp parallel
```
Comment (geogunow) [inline src/Cmfd.cpp]: Is this new spacing correct?

File: src/Cmfd.cpp [hunk 2]
```diff
@@ -1087,57 +1067,64 @@ void Cmfd::setNumFSRs(int num_fsrs) {
...
 
     FP_PRECISION current;
-    int cell_cw, cell_ccw;
-    int surface_cw, surface_ccw;
+    std::vector<int> surfaces;
+    std::vector<int>::iterator iter;
+    int cell, surface;
 
-    #pragma omp for
+#pragma omp for
```
Comment (geogunow) [inline src/Cmfd.cpp]: ditto

**Comment by geogunow** [comment]
I just read through this PR. Everything looks good aside from a couple small spacing comments. I have yet to run this on test cases. In the near future, this will hopefully be done by Travis.

File: src/Cmfd.cpp (hunk 1 above)
Comment (samuelshaner) [inline src/Cmfd.cpp]: All other preprocessor directives are indented to the line they address. However, the google c++ style guide says that all preprocessor directives should not be indented. This was accidental, but to be consistent with the style guide I would recommend we make a PR that removes the indent from the directives.

File: src/Cmfd.cpp [hunk 3]
```diff
@@ -1843,18 +1917,28 @@ void Cmfd::initialize() {
...
+  if (_cell_locks != NULL)
+    delete [] _cell_locks;
 
   try{
 
+    /* Allocate array of OpenMP locks for each CMFD cell */
+    _cell_locks = new omp_lock_t[num_cells];
+
+    /* Loop over all cells to initialize OpenMP locks */
+    #pragma omp parallel for schedule(guided)
```
Comment (geogunow) [inline src/Cmfd.cpp]: Per the new style, fix spacing

File: src/Vector.cpp [hunk 4]
```diff
@@ -82,6 +79,46 @@ void Vector::setAll(FP_PRECISION val) {
...
+               " which is not between 0 and %d", group_last, _num_groups-1);
+  else if (group_first > group_last)
+    log_printf(ERROR, "Unable to increment Vector values with first group %d"
+               " greater than last group %d", group_first, group_last);
+
+  /* Atomically increment the Vector values from the
+   * temporary array using mutual exclusion locks */
+  omp_set_lock(&_cell_locks[cell]);
+
+  #pragma omp simd
```
Comment (geogunow) [inline src/Vector.cpp]: Per the new style, fix spacing

File: src/Vector.cpp [hunk 5]
```diff
@@ -112,6 +149,46 @@ void Vector::setValue(int cell, int group, FP_PRECISION val) {
...
+               " which is not between 0 and %d", group_last, _num_groups-1);
+  else if (group_first > group_last)
+    log_printf(ERROR, "Unable to set Vector values with first group %d"
+               " greater than last group %d", group_first, group_last);
+
+  /* Atomically set the Vector values from the
+   * temporary array using mutual exclusion locks */
+  omp_set_lock(&_cell_locks[cell]);
+
+  #pragma omp simd
```
Comment (geogunow) [inline src/Vector.cpp]: Per the new style, fix spacing

File: src/linalg.cpp [hunk 6]
```diff
@@ -280,11 +282,12 @@ FP_PRECISION computeRMSE(Vector* X, Vector* Y, bool integrated) {
...
+  omp_lock_t* cell_locks = X->getCellLocks();
 
   if (integrated) {
 
     FP_PRECISION new_source, old_source;
-    Vector residual(num_x, num_y, 1);
+    Vector residual(cell_locks, num_x, num_y, 1);
 
     /* Compute the RMSE */
     #pragma omp parallel for private(new_source, old_source)
```
Comment (geogunow) [inline src/linalg.cpp]: Per the new style, fix spacing

File: src/Cmfd.cpp [hunk 7]
```diff
@@ -205,19 +205,19 @@ void Cmfd::setWidthY(double width) {
...
 void Cmfd::collapseXS() {
 
   log_printf(INFO, "Collapsing cross-sections onto CMFD mesh...");
 
-  /* Split corner currents to side surfaces */
-  splitCorners();
+  /* Split edge currents to side surfaces */
+  splitEdgeCurrents();
 
   #pragma omp parallel
```
Comment (geogunow) [inline src/Cmfd.cpp]: Per the new style, fix spacing

File: src/linalg.cpp [hunk 8]
```diff
@@ -304,7 +307,7 @@ FP_PRECISION computeRMSE(Vector* X, Vector* Y, bool integrated) {
   }
   else{
 
-    Vector residual(num_x, num_y, num_groups);
+    Vector residual(cell_locks, num_x, num_y, num_groups);
 
     /* Compute the RMSE */
     #pragma omp parallel for
```
Comment (geogunow) [inline src/linalg.cpp]: Per the new style, fix spacing

**Comment by geogunow** [comment]
Oops, forgot I merged this PR

**Pull request merged**